scripts/serialization/compare.py

```diff
@@ -0,0 +1,283 @@
+from shutil import rmtree
+from time import time
+from typing import Any, Callable, Dict, Iterator, List
+
+import numpy as np
+import pandas as pd
+from datasets import Dataset, disable_caching, load_dataset, load_from_disk  # pyright: ignore
+from matplotlib import pyplot as plt
+from tqdm import tqdm
+
+from streaming import MDSWriter, StreamingDataset
+
+disable_caching()
+
+
+def parse_args() -> Namespace:
+    """Parse command-line arguments.
+
+    Returns:
+        Namespace: Command-line arguments.
+    """
+    args = ArgumentParser()
+    args.add_argument('--data', type=str, default='/tmp/streaming/perf_test/')
+    args.add_argument('--num_samples', type=int, default=200_000)
+    args.add_argument('--out', type=str, default='plot.png')
+    return args.parse_args()
+
+
+def fetch_dataset(data_root: str, num_samples: int) -> List[Dict[str, Any]]:
+    """Fetch dataset to benchmark as JSONL.
+
+    Args:
+        data_root (str): Directory to cache samples in.
+        num_samples (int): Only use the first this many samples.
+
+    Returns:
+        List[Dict[str, Any]]: The samples.
+    """
+    if not os.path.isdir(data_root):
+        os.makedirs(data_root)
+    filename = os.path.join(data_root, 'samples.jsonl')
+
+    samples = []
+    if os.path.exists(filename):
+        for i, line in tqdm(enumerate(open(filename)), total=num_samples, leave=False):
+            if i == num_samples:
+                break
+            sample = json.loads(line)
+            samples.append(sample)
+
+        if len(samples) != num_samples:
+            samples = []
+            os.remove(filename)
+
+    if not os.path.exists(filename):
+        dataset = load_dataset('the_pile', 'pubmed_central', split='train', streaming=True)
+        with open(filename, 'w') as out:
+            for i, sample in tqdm(
+                    enumerate(dataset),
+                    total=len(dataset),  # pyright: ignore
+                    leave=False):
+                if i == num_samples:
+                    break
+                samples.append(sample)
+                line = json.dumps(sample, sort_keys=True) + '\n'
+                out.write(line)
+
+    return samples
+
+
+def arrow_write(samples: List[Dict[str, Any]], data_dir: str) -> None:
+    """Serialize dataset in arrow format.
+
+    Args:
+        samples (List[Dict[str, Any]]): List of sample dicts.
+        data_dir (str): Output data directory.
+    """
+    df = pd.DataFrame(samples)
+    ds = Dataset.from_pandas(df)
+    ds.save_to_disk(data_dir)
+
+
+def arrow_read_sequential(dirname: str) -> Iterator[Dict[str, Any]]:
+    """Iterate an arrow dataset sequentially.
+
+    Args:
+        dirname (str): Dataset directory.
+
+    Returns:
+        Iterator[Dict[str, Any]]: Iterator over samples.
+    """
+    dataset = load_from_disk(dirname)
+    dataset.remove_columns([col for col in dataset.column_names if col != 'text'])
+    yield from dataset.iter(1)  # pyright: ignore
```

Comment:
I think most people reading will know, but might be nice to be explicit here with `batch_size=1`